File: day16.py
from heapq import *
from collections import defaultdict

with open("data/input16.txt") as f:
    G=[list(x) for x in f.read().strip().split("\n")]
H,W=len(G),len(G[0])
sx,sy,ex,ey=0,0,0,0
for y in range(H):
    for x in range(W):
        if G[y][x]=="S": sx,sy=x,y
        if G[y][x]=="E": ex,ey=x,y

def neigh(x,y,h):
    nei = []
    for hh in {'L':['L','U','D'],'R':['R','U','D'],'U':['U','L','R'],'D':['D','L','R']}[h]:
        dx,dy={'L':(-1,0),'R':(1,0),'U':(0,-1),'D':(0,1)}[hh]
        nx,ny=x+dx,y+dy
        if G[ny][nx]!="#":
            nei.append((nx,ny,hh))
    return nei

seen = set()
dist=defaultdict(lambda:999999)
dist[(sx,sy,'R')]=0
Q = []
for y in range(H):
    for x in range(W):
        for h in ['L','R','D','U']:
            if G[y][x]=="#":
                continue
            if (x,y,h)==(sx,sy,'R'):heappush(Q,(0,(x,y,'R')))
            else: heappush(Q,(999999,(x,y,h)))
pred=defaultdict(list)               
while Q:
    d,u = heappop(Q)
    x,y,h=u
    if (x,y)==(ex,ey):
        break
    if (x,y,h) in seen:
        continue
    seen.add((x,y,h))
    for xx,yy,hh in neigh(x,y,h):
        if (xx,yy,hh) not in seen:
            if hh==h:
                deltaD=1
            else:
                deltaD=1001
            newdist = d+deltaD
            if newdist<=dist[(xx,yy,hh)]:
                dist[(xx,yy,hh)]=newdist
                heappush(Q,(newdist,(xx,yy,hh)))
                pred[(xx,yy,hh)].append((x,y,h))

def explore(x,y,h):
    G[y][x]='O'
    for xx,yy,hh in pred[(x,y,h)]:
        explore(xx,yy,hh)
            
best=min(dist[(ex,ey,h)] for h in ['L','R','U','D'])
for h in ['L','R','U','D']:
    if dist[(ex,ey,h)]==best:
        explore(ex,ey,h)

part1,part2=best,sum(line.count("O") for line in G)
print(part1,part2)

File: day16.lua
local part1,part2=0,0
local G={}
local sx,sy,W,H=0,0,0,0
for line in io.lines("data/input16.txt") do
    if #line>0 then
        H=H+1
        local row={}
        for i=1,#line do
            local ch=string.sub(line,i,i)
            table.insert(row,ch)
            if ch=="S" then sx,sy=i,H end
        end
        table.insert(G,row)
    end
end
W=#G[1]

function hash(x,y,h)
    return 4*(W*(y-1)+x-1)+h
end

local Q={{0,sx,sy,0}}
local seen={}
local dist={}
for y=1,H do
    for x=1,W do
        for h=0,3 do
            dist[hash(x,y,h)]=999999
        end
    end
end

local headings={{0,1,0},{1,0,1},{2,-1,0},{3,0,-1}}
local pred={}
while #Q>0 do
    local best,ibest=999999,1
    for i=1,#Q do
        local u=Q[i]
        if u[1]<best then best=u[1];ibest=i end
    end
    local u=table.remove(Q,ibest)
    local d,x,y,h=u[1],u[2],u[3],u[4]
    if G[y][x]=='E' then part1=dist[hash(x,y,h)];endstate=hash(x,y,h);break end
    if not seen[hash(x,y,h)] then
        seen[hash(x,y,h)]=true
        for _,H in ipairs(headings) do
            local hh,dx,dy=H[1],H[2],H[3]
            if G[y+dy][x+dx]~='#' and not seen[hash(x+dx,y+dy,hh)] then
                local deltaD = h==hh and 1 or 1001
                local newD=d+deltaD
                if dist[hash(x+dx,y+dy,hh)]>=newD then
                    dist[hash(x+dx,y+dy,hh)]=newD
                    table.insert(Q,{newD,x+dx,y+dy,hh})
                    local h1,h2=hash(x+dx,y+dy,hh),hash(x,y,h)
                    if pred[h1] then table.insert(pred[h1],h2) else pred[h1]={h2} end
                end
            end
        end
    end
end

tiles={}
function count(h)
    tiles[h//4]=true
    if not pred[h] then return end
    for _,hh in ipairs(pred[h]) do count(hh) end
end
for i=0,3 do
    count(4*(endstate//4)+i)
end
for _ in pairs(tiles) do part2=part2+1 end
print(part1,part2)